src/Expander.hs

```diff
@@ -926,28 +926,58 @@ runTask (tid, localData, task) = withLocal localData $ do
             addConstructor cn dt ctor ts
           linkDeclOutputScopes outScopesDest scs
     AwaitingPattern patPtr ty dest stx -> do
-      ready <-
-        case patPtr of
-          Left pptr -> isJust . view (expanderCompletedPatterns . at pptr) <$> getState
-          Right tptr -> isJust . view (expanderCompletedTypePatterns . at tptr) <$> getState
-      if (not ready)
+      ready <- do
+        let patternComplete ptr =
+              (view (expanderCompletedPatterns . at ptr) <$> getState) >>=
+              \case
+                Nothing ->
+                  pure False
+                Just layer ->
+                  and <$> traverse patternComplete layer
+        patternComplete patPtr
+        -- TODO traverse the tree and get the layers out, collecting the set of pointers to later accumulate the binders
+      if not ready
         then stillStuck tid task
         else do
-            varInfo <- view (expanderPatternBinders . at patPtr) <$> getState
-            case varInfo of
-              Nothing -> throwError $ InternalError "Pattern info not added"
-              Just vars -> do
-                p <- currentPhase
-                let rhs' = foldr (flip (addScope p)) stx
-                             [ sc'
-                             | (sc', _, _, _) <- vars
-                             ]
-                withLocalVarTypes
-                  [ (var, varStx, t)
-                  | (_sc, varStx, var, t) <- vars
-                  ] $
-                  expandOneExpression ty dest rhs'
-
+          let getVarInfo ptr =
+                (view (expanderPatternBinders . at ptr) <$> getState) >>=
+                \case
+                  Nothing ->
+                    throwError $ InternalError "Pattern info not added"
+                  Just (Right found) ->
+                    pure [found]
+                  Just (Left ptrs) ->
+                    concat <$> traverse getVarInfo ptrs
+          vars <- getVarInfo patPtr
+          p <- currentPhase
+          let rhs' = foldr (flip (addScope p)) stx
+                       [ sc'
+                       | (sc', _, _, _) <- vars
+                       ]
+          withLocalVarTypes
+            [ (var, varStx, t)
+            | (_sc, varStx, var, t) <- vars
+            ] $
+            expandOneExpression ty dest rhs'
+    AwaitingTypePattern patPtr ty dest stx -> do
+      ready <- isJust . view (expanderCompletedTypePatterns . at patPtr) <$> getState
```

Comment:
a slightly more lensy way of expressing this would be

    ready <- has (expanderCompletedTypePatterns . ix patPtr) <$> getState

Reply:
Is lensiness better here? I'm trying to use lenses in a way that don't make the code incomprehensible to those who haven't marinated their brains in it :)

Reply:
lensy idioms often work better with nested structures; for example, `has (myMap1 . ix k1 . myMap2 . ix k2) s1` seems better to me than

    case view (myMap1 . at k1) s1 of
      Nothing -> False
      Just s2 -> isJust $ view (myMap2 . at k2) s2

but since we're not dealing with a deep structure here, it doesn't really matter. besides, I wasn't really trying to suggest an improvement, I was only trying to spread my knowledge of lensy tricks, in the hope you find that information useful!

Reply:
That makes sense! Thanks!

Reply:
Does this mean that this version of `pruneLevel` would be equivalent?
```
pruneLevel :: Traversable f => BindingLevel -> f MetaPtr -> Expand ()
pruneLevel l = traverse_ reduce
  where
    reduce ptr =
      modifyState $
      over (expanderTypeStore . ix ptr . varLevel) (min l)
```
The current one is:
```
pruneLevel :: Traversable f => BindingLevel -> f MetaPtr -> Expand ()
pruneLevel l = traverse_ reduce
  where
    reduce ptr =
      modifyState $
      over (expanderTypeStore . at ptr) $
      fmap (over varLevel (min l))
```

Reply:
yes!